spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java

```diff
@@ -106,37 +110,59 @@ else if (!StringUtils.hasText(this.defaultBinder)) {
 					}
 				}
 				if (defaultCandidateConfigurations.size() == 1) {
-					this.defaultBinder = defaultCandidateConfigurations.iterator().next();
-					configurationName = this.defaultBinder;
+					this.defaultBindersPerBoundType.put(boundElementType.getName(),
+							defaultCandidateConfigurations.iterator().next());
+					configurationName = this.defaultBindersPerBoundType.get(boundElementType.getName());
 				}
 				else {
 					if (defaultCandidateConfigurations.size() > 1) {
-						throw new IllegalStateException(
-								"A default binder has been requested, but there is more than one binder available: "
-										+ StringUtils.collectionToCommaDelimitedString(defaultCandidateConfigurations)
-										+ ", and no default binder has been set.");
+						List<String> candidatesForBindableType = new ArrayList<>();
+						for (String defaultCandidateConfiguration : defaultCandidateConfigurations) {
+							Binder<Object, ?, ?> binderInstance = getBinderInstance(defaultCandidateConfiguration);
+							Class<?> binderType = GenericsUtils.getParameter(binderInstance.getClass(), Binder.class, 0);
+							if (binderType.isAssignableFrom(boundElementType)) {
+								candidatesForBindableType.add(defaultCandidateConfiguration);
+							}
+						}
+						if (candidatesForBindableType.size() == 1) {
+							this.defaultBindersPerBoundType.put(boundElementType.getName(),
+									candidatesForBindableType.iterator().next());
+							configurationName = this.defaultBindersPerBoundType.get(boundElementType.getName());
+						} else if (candidatesForBindableType.size() > 1) {
+							throw new IllegalStateException(
+									"A default binder has been requested, but there is more than one binder available for '"
+											+ boundElementType.getName() + "' : "
+											+ StringUtils.collectionToCommaDelimitedString(defaultCandidateConfigurations)
```

Comment:
I think this one should only print out the `candidatesForBindableType` collection rather than the full `defaultCandidateConfigurations` collection.